### PR TITLE
BUG: Presence of softlink in HDF5 file breaks HDFStore.keys() (GH20523)

### LINKED ISSUE #20523: Presence of softlink in HDF5 file breaks HDFStore.keys()
#### Code Sample, a copy-pastable example if possible

```python
#! /path/to/python3.6

import pandas as pd

df = pd.DataFrame({ "a": [1], "b": [2] })
print(df.to_string())

hdf = pd.HDFStore("/tmp/test.hdf", mode="w")
hdf.put("/test/key", df)

#Brittle
hdf._handle.create_soft_link(hdf._handle.root.test, "symlink", "/test/key")
hdf.close()
print("Successful write")

hdf = pd.HDFStore("/tmp/test.hdf", mode="r")
'''
Traceback (most recent call last):
  File "snippet.py", line 31, in <module>
    print(hdf.keys())
  File "python3.6.3/lib/python3.6/site-packages/pandas/io/pytables.py", line 529, in keys
    return [n._v_pathname for n in self.groups()]
  File "python3.6.3/lib/python3.6/site-packages/pandas/io/pytables.py", line 1077, in groups
    g for g in self._handle.walk_nodes()
  File "python3.6.3/lib/python3.6/site-packages/pandas/io/pytables.py", line 1078, in <listcomp>
    if (getattr(g._v_attrs, 'pandas_type', None) or
  File "python3.6.3/lib/python3.6/site-packages/tables/link.py", line 79, in __getattr__
    "`%s` instance" % self.__class__.__name__)
KeyError: 'you cannot get attributes from this `NoAttrs` instance'
'''
print(hdf.keys()) #causes exception
hdf.close()

print("Successful read")

```

#### Problem description

I know I have a esoteric problem, but I'm building an HDF5 file using Pandas and then using pytables to softlink to the Pandas dataframe. I understand this is unsupported and brittle but for my use case I haven't been able to come up with a better/simpler solution.

This issue is similar to: https://github.com/pandas-dev/pandas/issues/6019

The root cause is when we call HDFStore.keys(), it calls HDFStore.groups() and eventually g._v_attrs on a Pytables File.

https://github.com/pandas-dev/pandas/blob/master/pandas/io/pytables.py#L1076

But calling g._v_attrs on a tables.link.SoftLink causes a KeyError due to:

https://github.com/PyTables/PyTables/blob/develop/tables/link.py#L76

And there doesn't look to be a way to guard against an instance of NoAttrs since that class is defined within the method. One solution may be to check the instance of g if it's a Link
```
        return [
            g for g in self._handle.walk_nodes()
            if (not isinstance(g, _table_mod.link.Link) and
                (getattr(g._v_attrs, 'pandas_type', None) or
                 getattr(g, 'table', None) or
                (isinstance(g, _table_mod.table.Table) and
                 g._v_name != u('table'))))
        ]
```

I'd be happy to write a PR and tests if you find this change acceptable.

#### Expected Output

```
   a  b
0  1  2
Successful write
['/test/key']
Successful read
```

#### Output of ``pd.show_versions()``

<details>
                                                                                                                                                                                                                              
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.3.final.0
python-bits: 64
OS: Linux
OS-release: 3.10.0-514.21.1.el7.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: en_US.utf-8
LANG: en_US.utf-8
LOCALE: en_US.UTF-8

pandas: 0.22.0
pytest: None
pip: 9.0.1
setuptools: 38.5.1
Cython: None
numpy: 1.14.0
scipy: 1.0.0
pyarrow: None
xarray: None
IPython: 6.2.1
sphinx: None
patsy: 0.4.1
dateutil: 2.6.1
pytz: 2018.3
blosc: None
bottleneck: None
tables: 3.4.2
numexpr: 2.6.4
feather: None
matplotlib: 2.1.0
openpyxl: None
xlrd: 1.1.0
xlwt: None
xlsxwriter: None
lxml: None
bs4: 4.6.0
html5lib: 1.0.1
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by dworvos**
- [x] closes #20523 
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/io/pytables.py [hunk 1]
```diff
@@ -1073,10 +1073,11 @@ def groups(self):
         self._check_if_open()
         return [
             g for g in self._handle.walk_nodes()
-            if (getattr(g._v_attrs, 'pandas_type', None) or
-                getattr(g, 'table', None) or
+            if (not isinstance(g, _table_mod.link.Link) and
```
Comment (TomAugspurger) [inline pandas/io/pytables.py]: Are there any other cases we should exclude, or just Link?

Comment (dworvos) [reply pandas/io/pytables.py]: Link is the abstract base class for SoftLink and ExternalLink in pytables. These subclasses do not support attributes as of HDF5 (this appears to be a limitation of the HDF5 library but may change in the future, reference: http://www.pytables.org/usersguide/libref/link_classes.html). I believe all other node types support attributes. A quick search of NoAttrs in Pytables shows usage only in tables/link.py.

**Review by TomAugspurger — APPROVED** [review]

File: pandas/tests/io/test_pytables.py [hunk 2]
```diff
@@ -5521,3 +5521,20 @@ def test_dst_transitions(self):
                 store.append('df', df)
                 result = store.select('df')
                 assert_frame_equal(result, df)
+
+
+def test_ignore_hdf_softlink():
+
+    with ensure_clean_store('tmp.__%s__.h5' % tm.rands(10)) as store:
+
+        # GH 20523
```
Comment (jreback) [inline pandas/tests/io/test_pytables.py]: can you move this test right after test_keys (in the class)

File: pandas/tests/io/test_pytables.py [hunk 3]
```diff
@@ -5521,3 +5521,20 @@ def test_dst_transitions(self):
                 store.append('df', df)
                 result = store.select('df')
                 assert_frame_equal(result, df)
+
+
+def test_ignore_hdf_softlink():
+
+    with ensure_clean_store('tmp.__%s__.h5' % tm.rands(10)) as store:
```
Comment (jreback) [inline pandas/tests/io/test_pytables.py]: move the comment right below the test title

File: doc/source/whatsnew/v0.23.0.txt [hunk 4]
```diff
@@ -1020,6 +1020,7 @@ I/O
 - Bug in :meth:`pandas.io.stata.StataReader.value_labels` raising an ``AttributeError`` when called on very old files. Now returns an empty dict (:issue:`19417`)
 - Bug in :func:`read_pickle` when unpickling objects with :class:`TimedeltaIndex` or :class:`Float64Index` created with pandas prior to version 0.20 (:issue:`19939`)
 - Bug in :meth:`pandas.io.json.json_normalize` where subrecords are not properly normalized if any subrecords values are NoneType (:issue:`20030`)
+- Bug in :func:`HDFStore.groups` when reading a file with a softlink causes exception (:issue:`20523`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.0.txt]: say ``.keys`` instead here as that's more user facing.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by dworvos** [comment]
Hi, had some newbie git issues yesterday with rebasing, but I think I resolved them now in my branch/fork - please let me know if there are outstanding issues.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @dworvos

**Pull request merged**